crates/subspace-archiving/src/archiver/incremental_record_commitments.rs

```diff
@@ -0,0 +1,194 @@
+extern crate alloc;
+
+use crate::archiver::Segment;
+use alloc::collections::VecDeque;
+use blake2::digest::typenum::U32;
+use blake2::digest::{FixedOutput, Update};
+use blake2::Blake2b;
+use core::mem;
+use parity_scale_codec::{Encode, Output};
+use subspace_core_primitives::Blake2b256Hash;
+
+/// State of incremental record commitments, encapsulated to hide implementation details and
+/// encapsulate tricky logic
+#[derive(Debug, Default, Clone)]
+pub(super) struct IncrementalRecordCommitmentsState {
+    /// State contains record commitments.
+    ///
+    /// NOTE: Until full segment is processed, this will not contain commitment to the first record
+    /// since it is not ready yet. This in turn means all commitments will be at `-1` offset.
+    state: VecDeque<Blake2b256Hash>,
+}
+
+impl IncrementalRecordCommitmentsState {
+    /// Creates an empty state with space for at least capacity records.
+    pub(super) fn with_capacity(capacity: usize) -> Self {
+        Self {
+            state: VecDeque::with_capacity(capacity),
+        }
+    }
+
+    pub(super) fn drain(&mut self) -> impl Iterator<Item = Blake2b256Hash> + '_ {
+        self.state.drain(..)
+    }
+}
+
+/// Update internal record commitments state based on (full or partial) segment.
+pub(super) fn update_record_commitments(
+    incremental_record_commitments: &mut IncrementalRecordCommitmentsState,
+    segment: &Segment,
+    full: bool,
+    record_size: usize,
+) {
+    segment.encode_to(&mut IncrementalRecordCommitmentsProcessor::new(
+        incremental_record_commitments,
+        record_size,
+        full,
+    ));
+}
+
+/// Processor is hidden to not expose unnecessary implementation details (like `Output` trait
+/// implementation)
+struct IncrementalRecordCommitmentsProcessor<'a> {
+    /// Processed bytes in the segment so far
+    processed: usize,
+    /// Record commitments already created
+    incremental_record_commitments: &'a mut IncrementalRecordCommitmentsState,
+    /// Record size
+    record_size: usize,
+    /// Whether segment is full or partial
+    full: bool,
+    /// Intermediate hashing state that computes Blake2-256-254.
+    ///
+    /// See [`subspace_core_primitives::crypto::blake2b_256_254_hash`] for details.
+    hashing_state: Blake2b<U32>,
+}
+
+impl<'a> Drop for IncrementalRecordCommitmentsProcessor<'a> {
+    fn drop(&mut self) {
+        if self.full {
+            // How many bytes to read before the start of the next record
+            let bytes_until_full_record = self.bytes_until_full_record();
+            if bytes_until_full_record > 0 {
+                // This is fine since we'll have at most a few iterations and allocation is less
+                // desirable than a loop here
+                for _ in 0..bytes_until_full_record {
+                    self.update_hashing_state(&[0]);
```

Comment:
Why do we pass zero here?

Reply:
We're pushing a few zero bytes as padding til the end of the last record of the segment.